core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java

```diff
@@ -285,26 +286,38 @@ private void matchRecurse(int solve) {
         final int parentOrdinal = operand.getParent().ordinalInRule;
         final RelNode parentRel = rels[parentOrdinal];
         final List<RelNode> inputs = parentRel.getInputs();
-        if (operand.ordinalInParent < inputs.size()) {
-          final RelSubset subset =
-              (RelSubset) inputs.get(operand.ordinalInParent);
-          if (operand.getMatchedClass() == RelSubset.class) {
-            successors = subset.set.subsets;
-          } else {
-            successors = subset.getRelList();
+        // if the child is unordered, then add all rels in all subsets to the successor list
+        // unordered can match child in any ordinal
+        if (parentOperand.childPolicy.equals(RelOptRuleOperandChildPolicy.UNORDERED)) {
```

Comment:
This looks like the most interesting part of the change, however I am not familiar with this part, so it is hard for me to review it :-/